src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -241,77 +251,99 @@ public ValueTask<Variant> GetVariantAsync(string feature, TargetingContext conte
                 throw new ArgumentNullException(nameof(context));
             }
 
-            return GetVariantAsync(feature, context, useContext: true, cancellationToken);
+            EvaluationEvent evaluationEvent = await EvaluateFeature(feature, context, useContext: true, cancellationToken);
+
+            return evaluationEvent.Variant;
         }
 
-        private async ValueTask<bool> IsEnabledWithVariantsAsync<TContext>(string feature, TContext appContext, bool useAppContext, CancellationToken cancellationToken)
+        private async Task<EvaluationEvent> EvaluateFeature<TContext>(string feature, TContext context, bool useContext, CancellationToken cancellationToken)
         {
-            bool isFeatureEnabled = false;
-
-            FeatureDefinition featureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false);
-
-            VariantDefinition variantDefinition = null;
+            var evaluationEvent = new EvaluationEvent
+            {
+                FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
+            };
 
-            if (featureDefinition != null)
+            if (evaluationEvent.FeatureDefinition != null)
             {
-                isFeatureEnabled = await IsEnabledAsync(featureDefinition, appContext, useAppContext, cancellationToken).ConfigureAwait(false);
+                //
+                // Determine IsEnabled
+                evaluationEvent.Enabled = await IsEnabledAsync(evaluationEvent.FeatureDefinition, context, useContext, cancellationToken).ConfigureAwait(false);
+
+                //
+                // Determine Variant
+                VariantDefinition variantDefinition = null;
 
-                if (featureDefinition.Variants != null && featureDefinition.Variants.Any() && featureDefinition.Allocation != null)
+                if (evaluationEvent.FeatureDefinition.Allocation != null && (evaluationEvent.FeatureDefinition.Variants?.Any() ?? false))
```

Comment:
If there is at least one variant then I believe we should run assignment. In the case that there is at least one variant, but allocation is null then I expect the variant returned should be `null` but the assignment reason should be `DefaultWhenDisabled`. Allocation being null in this case is the same as Allocation being not null, but `Allocation.DefaultWhenDisabled` being null.

Reply:
Can invert the logic below to check for 

```
if (!evaluationEvent.Enabled ||
    evaluationEvent.Allocation == null)```

Reply:
To make it clear, I put the assignment process here:
![image](https://github.com/microsoft/FeatureManagement-Dotnet/assets/141655842/6a8db7a9-0bfe-4de5-a7eb-941c0aab7f5f)

> In the case that there is at least one variant, but allocation is null then I expect the variant returned should be null but the assignment reason should be DefaultWhenDisabled. Allocation being null in this case is the same as Allocation being not null, but Allocation.DefaultWhenDisabled being null.

I am a bit confused, why should the reason be `DefaultWhenDisabled`. I think the reason should be `DefaultWhenEnabled` or `DefaultWhenDisabled` depending on whether the feature flag is on or off.

Reply:
Discussed offline, I should have said it is either DefualtWhenDisabled or DefaultWhenEnabled